### DISCUSSION TITLE
Extending the scene-graph API - Model manipulation

### DISCUSSION THREAD

**Discussion by felipe-chamas** [google/model-viewer#1817] [Q&A] [answered]
I'm currently interested in real-time 3Dmodel web manipulation and because we are using `model-viewer` we wish on staying with it as our Model Viewer 🙃 framework. Here is a list of what we might need:

- Possible in MV ✔️ 
  - Change a whole model
  - Manipulate materials on a model 
- Not possible ❌ (yet 🤞) 
  - Manipulate parts of a model
    - Hide/Show children of the GLTF tree
    - Replace children
  - Load more than one src in a single viewer
  - Manipulate materials of each loaded src

Studying the code and the [question](https://spectrum.chat/model-viewer/general/scene-graph-api-change-objects-in-a-scene~29400562-a4b1-4c0b-a420-b360dc106275) made by @aviggiano I wondered about the starting point on implementing this idea. Since it deals with lower level manipulation, I need a little help understanding if it's better learning three.js, 3DOM or if it's easier to use/extend the already implemented API somehow.
If I can be in any help clarifying our needs, please let me know, and I hope we can contribute with these modifications through a PR once we manage to work it out.

**Comment by elalish** [comment, marked as answer]
Fantastic! These are definitely capabilities we want to enable with the scene-graph API and help would be much appreciated. You'll notice 3DOM is no longer its own package, which is because we simplified it considerably and put all the logic into the scene-graph mixin, which should make this work easier. I would recommend reviewing the three.js GLTFLoader in particular and how we integrate that into our scene-graph API. Currently we only expose materials, so this will involve exposing nodes as well. The main thing is to ensure our API is based on the GLTF itself and abstracts the specifics of the three.js implementation. 

You might want to start a PR with just a design doc and we can do review and have back-and-forth there before you start implementing.

**Reply by felipe-chamas** [reply]
Amazing, thanks for the heads up! I'm gonna start my studies in it and begin the preparation for this roadmap of design doc!

**Reply by simonbarker** [reply]
Did this ever go anywhere? I'm looking at moving to model-viewer from Sketchfab and need this exact capability. The SketchFab viewer API allows you to search the tree and then show and hide based on node id.

Keen to know if this progressed before potentially diving in myself.

**Reply by elalish** [reply]
We don't expose nodes, but we do enable show/hide of materials: https://modelviewer.dev/examples/scenegraph/#hideShowMeshVariantsExample

So as long as you author each thing you want to hide with its own material, you can get them by name. Does that solve your use case?

**Reply by simonbarker** [reply]
Thanks for the reply. Sadly not, but I worked out if I import ThreeJS separately then I can find the full node tree via the .model property, recursively search the tree looking for our named nodes then toggle .visible and force a re-render. So I think I can do what I need.

I’m looking to move away from Sketchfab and need to reimplement much of their ViewerAPI functionality.
